luigi/contrib/sge.py

```diff
@@ -0,0 +1,317 @@
+shared cluster. Jobs are submitted using the ``qsub`` command and monitored
+using ``qstat``. To get started, install luigi on all nodes.
+
+
+To run luigi workflows on an SGE cluster, subclass
+:class:`luigi.contrib.sge.SGEJobTask` as you would any :class:`luigi.Task`,
+but override the ``work()`` method, instead of ``run()``, to define the job
+code. Then, run your Luigi workflow from the master node, assigning > 1
+``workers`` in order to distribute the tasks in parallel across the cluster.
+
+A few SGE global parameters can be added to the luigi configuration. Below is
+an example, showing the defaults (so if this looks ok then you don't need to
+change the config). Defaults were matched to the values used by StarCluster,
+an open-source SGE cluster manager for use with Amazon EC2. The ``n-cpu``
+configuration is also a Parameter in :class:`luigi.contrib.sge.SGEJobTask`,
+allowing you to define multiple tasks, each with different compute resource
+requirements (or slots, in SGE terms)::
+
+    [sge]
+    shared-tmp-dir = /home
+    parallel-env = orte
+    n-cpu = 2
+
+
+The following is an example usage (and can also be found in ``sge_tests.py``)::
+
+    import logging
+    import luigi
+    from luigi.contrib.sge import SGEJobTask
+
+    logger = logging.getLogger('luigi-interface')
+
+
+    class TestJobTask(SGEJobTask):
+
+        i = luigi.Parameter()
+
+        def work(self):
+            logger.info('Running test job...')
+            with open(self.output().path, 'w') as f:
+                f.write('this is a test')
+
+        def output(self):
+            return luigi.LocalTarget(os.path.join('/home', 'testfile_' + str(self.i)))
+
+
+    if __name__ == '__main__':
+        tasks = [TestJobTask(i=str(i), n_cpu=i+1) for i in range(3)]
+        luigi.build(tasks, local_scheduler=True, workers=3)
+
+
+In this example, the third Task asks for 3 CPU slots. If your cluster only
+contains nodes with 1 or 2 CPUs, this task will hang indefinitely in the queue.
+
+"""
+
+
+# This extension is modeled after the hadoop.py approach.
+#
+# Implementation notes
+# The procedure:
+# - Pickle the class
+# - Construct a qsub argument that runs a generic runner function with the path to the pickled class
+# - Runner function loads the class from pickle
+# - Runner function hits the work button on it
+
+import os
+import subprocess
+import time
+import sys
+import logging
+import random
+import shutil
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+import luigi
+import luigi.hadoop
+from luigi import configuration
+from luigi.contrib import sge_runner
+
+logger = logging.getLogger('luigi-interface')
+logger.propagate = 0
+
+POLL_TIME = 5  # decided to hard-code rather than configure here
+
+# Defaults for configuration parameters. Use StarCluster defaults. Number of CPUs is further configurable as task parameter
+DEFAULT_HOME = configuration.get_config().get('sge', 'shared-tmp-dir', default=None) or '/home'
```

Comment:
Can you convert all configuration to the "new way" with parameters? Grep `luigi.Config` for examples :)